yoyodyne/models/hmm.py

```diff
@@ -0,0 +1,413 @@
+        log_probs, alignments = self(batch)
+        loss = self.loss_func(batch.target.padded, log_probs, alignments)
+        self.log(
+            "train_loss",
+            loss,
+            batch_size=len(batch),
+            on_step=False,
+            on_epoch=True,
+        )
+        return loss
+
+    def validation_step(self, batch: data.PaddedBatch, batch_idx: int) -> Dict:
+        predictions, likelihood = self(batch)
+        # Processes for accuracy calculation.
+        predictions = self.evaluator.finalize_predictions(
+            predictions, self.end_idx, self.pad_idx
+        )
+        val_eval_item = self.evaluator.get_eval_item(
+            predictions, batch.target.padded, self.pad_idx
+        )
+        return {"val_eval_item": val_eval_item, "val_loss": -likelihood.mean()}
+
+    def predict_step(self, batch: data.PaddedBatch, batch_idx: int) -> Dict:
+        predictions, _ = self(batch)
+        # Processes for accuracy calculation.
+        predictions = self.evaluator.finalize_predictions(
+            predictions, self.end_idx, self.pad_idx
+        )
+        return predictions
+
+    def get_decoder(self) -> modules.lstm.HMMLSTMDecoder:
+        if self.attention_context > 0:
+            return modules.lstm.ContextHMMLSTMDecoder(
+                pad_idx=self.pad_idx,
+                start_idx=self.start_idx,
+                end_idx=self.end_idx,
+                attention_context=self.attention_context,
+                decoder_input_size=(
+                    self.source_encoder.output_size
+                    + self.features_encoder.output_size
+                    if self.has_features_encoder
+                    else self.source_encoder.output_size
+                ),
+                num_embeddings=self.target_vocab_size,
+                dropout=self.dropout,
+                bidirectional=False,
+                embedding_size=self.embedding_size,
+                layers=self.decoder_layers,
+                hidden_size=self.hidden_size,
+            )
+        else:
+            return modules.lstm.HMMLSTMDecoder(
+                pad_idx=self.pad_idx,
+                start_idx=self.start_idx,
+                end_idx=self.end_idx,
+                decoder_input_size=(
+                    self.source_encoder.output_size
+                    + self.features_encoder.output_size
+                    if self.has_features_encoder
+                    else self.source_encoder.output_size
+                ),
+                num_embeddings=self.target_vocab_size,
+                dropout=self.dropout,
+                bidirectional=False,
+                embedding_size=self.embedding_size,
+                layers=self.decoder_layers,
+                hidden_size=self.hidden_size,
+            )
+
+    def _get_loss_func(
+        self,
+    ) -> Callable[[torch.Tensor, torch.Tensor, torch.Tensor], torch.Tensor]:
+        # TODO: Make the loss update continuous during training to avoid.
```

Comment:
I don't understand this TODO because I don't know what you're avoiding.

Reply:
This method is here to keep with the base model interface, it should only be called once when the model is initialized and just tells the model what to use as `self.loss_func`. Since you override all that stuff I guess you don't need it, but I don't see why removing it would benefit you?

Reply:
(If you keep this) just move explanation into TODO to increase bus factor

Reply:
@Adamits Oh, I still need to keep the _get_loss_func. My point is that at the moment the model builds a psssh, len x len x vocab_size x num_steps size tensor to track emissions and transitions. I need to simplify this.

@kylebgorman ???

Reply:
gotcha, i moved it to reduce confusion. wasn't getting the bus factor reference.